moveit_experimental/moveit_jog_arm/src/collision_check_thread.cpp

```diff
@@ -46,108 +49,139 @@ namespace moveit_jog_arm
+
   if (parameters_.collision_check_rate < MIN_RECOMMENDED_COLLISION_RATE)
     ROS_WARN_STREAM_THROTTLE_NAMED(5, LOGNAME, "Collision check rate is low, increase it in yaml file if CPU allows");
+
+  // subscribe to joints
+  joint_state_sub_ = nh_.subscribe(parameters.joint_topic, 1, &CollisionCheckThread::jointStateCB, this);
+
+  // Publish to internal namespace
+  ros::NodeHandle internal_nh("~internal");
+  collision_velocity_scale_pub_ = internal_nh.advertise<std_msgs::Float64>("collision_velocity_scale", 1);
+
+  // Wait for incoming topics to appear
+  ROS_DEBUG_NAMED(LOGNAME, "Waiting for JointState topic");
+  ros::topic::waitForMessage<sensor_msgs::JointState>(parameters_.joint_topic);
+
+  current_state_ = std::make_unique<moveit::core::RobotState>(getLockedPlanningSceneRO()->getCurrentState());
+  acm_ = getLockedPlanningSceneRO()->getAllowedCollisionMatrix();
 }
 
 planning_scene_monitor::LockedPlanningSceneRO CollisionCheckThread::getLockedPlanningSceneRO() const
 {
   return planning_scene_monitor::LockedPlanningSceneRO(planning_scene_monitor_);
 }
 
-void CollisionCheckThread::startMainLoop(JogArmShared& shared_variables)
+void CollisionCheckThread::start()
 {
-  // Init collision request
-  collision_detection::CollisionRequest collision_request;
-  collision_request.group_name = parameters_.move_group_name;
-  collision_request.distance = true;  // enable distance-based collision checking
-  collision_detection::CollisionResult collision_result;
-
-  // Copy the planning scene's version of current state into new memory
-  moveit::core::RobotState current_state(getLockedPlanningSceneRO()->getCurrentState());
-
-  const double self_velocity_scale_coefficient = -log(0.001) / parameters_.self_collision_proximity_threshold;
-  const double scene_velocity_scale_coefficient = -log(0.001) / parameters_.scene_collision_proximity_threshold;
-  ros::Rate collision_rate(parameters_.collision_check_rate);
-
-  double self_collision_distance = 0;
-  double scene_collision_distance = 0;
-  bool collision_detected;
-
-  // Scale robot velocity according to collision proximity and user-defined thresholds.
-  // I scaled exponentially (cubic power) so velocity drops off quickly after the threshold.
-  // Proximity decreasing --> decelerate
-  double velocity_scale = 1;
-
-  collision_detection::AllowedCollisionMatrix acm = getLockedPlanningSceneRO()->getAllowedCollisionMatrix();
-  /////////////////////////////////////////////////
-  // Spin while checking collisions
-  /////////////////////////////////////////////////
-  while (ros::ok() && !shared_variables.stop_requested)
+  timer_ = nh_.createTimer(period_, &CollisionCheckThread::run, this);
+}
+
+void CollisionCheckThread::stop()
+{
+  timer_.stop();
+}
+
+void CollisionCheckThread::run(const ros::TimerEvent& timer_event)
+{
+  // Log last loop duration and warn if it was longer than period
+  if (timer_event.profile.last_duration.toSec() < period_.toSec())
```

Comment:
One thing I could change is publish these durations to an internal topic that is never subscribed to.  That would be almost free and would allow introspection through a bag if you cared (although that would result in a performance hit for measuring performance).  The downsides to that approach could be mitigated by @gavanderhoorn's suggestion down below about using nodelet_bag (if jog_arm was a nodelet itself).